File: week1/92.reverse_linked_list_ii-反转链表2/92.reverse_linked_list_ii.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseBetween(ListNode head, int left, int right) {
        ListNode dummy = new ListNode();
        dummy.next = head;

        ListNode guard = dummy;
        ListNode point = dummy.next;

        for (int step = 0; step < left - 1; step++) {
            guard = guard.next;
            point = point.next;
        }

        for (int i = 0; i < right - left; i++) {
            ListNode removed = point.next;
            point.next = point.next.next;
            removed.next = guard.next;
            guard.next = removed;
        }
        return dummy.next;
    }
}

File: Leet/92.reverse_linked_list_ii-反转链表2/92.reverse_linked_list_ii.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} left
 * @param {number} right
 * @return {ListNode}
 */
var reverseBetween = function (head, left, right) {
    let dummy = new ListNode();
    dummy.next = head;
    let g = dummy, p = dummy.next;

    for (let step = 0; step < left - 1; step++) {
        g = g.next; p = p.next;
    }

    for (let i = 0; i < right - left; i++) {
        let removed = p.next;
        p.next = p.next.next;
        removed.next = g.next;
        g.next = removed;
    }
    return dummy.next;
};